crypto/fipsmodule/ec/p384.c

```diff
@@ -0,0 +1,452 @@
+  fiat_p384_sub(y_out, y_out, gamma);
+}
+
+// fiat_p384_point_add calculates (x1, y1, z1) + (x2, y2, z2)
+//
+// The method is taken from:
+//   http://hyperelliptic.org/EFD/g1p/auto-shortw-jacobian.html#addition-add-2007-bl
+// adapted for mixed addition (z2 = 1, or z2 = 0 for the point at infinity).
+//
+// Coq transcription and correctness proof:
+// <https://github.com/davidben/fiat-crypto/blob/c7b95f62b2a54b559522573310e9b487327d219a/src/Curves/Weierstrass/Jacobian.v#L467>
+// <https://github.com/davidben/fiat-crypto/blob/c7b95f62b2a54b559522573310e9b487327d219a/src/Curves/Weierstrass/Jacobian.v#L544>
+static void fiat_p384_point_add(fiat_p384_felem x3, fiat_p384_felem y3,
+                                fiat_p384_felem z3, const fiat_p384_felem x1,
+                                const fiat_p384_felem y1,
+                                const fiat_p384_felem z1, const int mixed,
+                                const fiat_p384_felem x2,
+                                const fiat_p384_felem y2,
+                                const fiat_p384_felem z2) {
+  fiat_p384_felem x_out, y_out, z_out;
+  fiat_p384_limb_t z1nz = fiat_p384_nz(z1);
+  fiat_p384_limb_t z2nz = fiat_p384_nz(z2);
+
+  // z1z1 = z1z1 = z1**2
+  fiat_p384_felem z1z1;
+  fiat_p384_square(z1z1, z1);
+
+  fiat_p384_felem u1, s1, two_z1z2;
+  if (!mixed) {
+    // z2z2 = z2**2
+    fiat_p384_felem z2z2;
+    fiat_p384_square(z2z2, z2);
+
+    // u1 = x1*z2z2
+    fiat_p384_mul(u1, x1, z2z2);
+
+    // two_z1z2 = (z1 + z2)**2 - (z1z1 + z2z2) = 2z1z2
+    fiat_p384_add(two_z1z2, z1, z2);
+    fiat_p384_square(two_z1z2, two_z1z2);
+    fiat_p384_sub(two_z1z2, two_z1z2, z1z1);
+    fiat_p384_sub(two_z1z2, two_z1z2, z2z2);
+
+    // s1 = y1 * z2**3
+    fiat_p384_mul(s1, z2, z2z2);
+    fiat_p384_mul(s1, s1, y1);
+  } else {
+    // We'll assume z2 = 1 (special case z2 = 0 is handled later).
+
+    // u1 = x1*z2z2
+    fiat_p384_copy(u1, x1);
+    // two_z1z2 = 2z1z2
+    fiat_p384_add(two_z1z2, z1, z1);
+    // s1 = y1 * z2**3
+    fiat_p384_copy(s1, y1);
+  }
+
+  // u2 = x2*z1z1
+  fiat_p384_felem u2;
+  fiat_p384_mul(u2, x2, z1z1);
+
+  // h = u2 - u1
+  fiat_p384_felem h;
+  fiat_p384_sub(h, u2, u1);
+
+  fiat_p384_limb_t xneq = fiat_p384_nz(h);
+
+  // z_out = two_z1z2 * h
+  fiat_p384_mul(z_out, h, two_z1z2);
+
+  // z1z1z1 = z1 * z1z1
+  fiat_p384_felem z1z1z1;
+  fiat_p384_mul(z1z1z1, z1, z1z1);
+
+  // s2 = y2 * z1**3
+  fiat_p384_felem s2;
+  fiat_p384_mul(s2, y2, z1z1z1);
+
+  // r = (s2 - s1)*2
+  fiat_p384_felem r;
+  fiat_p384_sub(r, s2, s1);
+  fiat_p384_add(r, r, r);
+
+  fiat_p384_limb_t yneq = fiat_p384_nz(r);
+
+  // This case will never occur in the constant-time |ec_GFp_mont_mul|.
+  fiat_p384_limb_t is_nontrivial_double = constant_time_is_zero_w(xneq | yneq) &
+                                          ~constant_time_is_zero_w(z1nz) &
+                                          ~constant_time_is_zero_w(z2nz);
+  if (is_nontrivial_double) {
```

Comment:
you're right. This handling is caught in special tests. It happens rarely, if at all. It says in the comment above it would not occur. We need to check in our implementations of scalar multiplication that it will not occur.

Reply:
I'll try to remember.